R/pivot.R

```diff
@@ -0,0 +1,139 @@
+#'
+#' @param dataset Built dataset with `test_build_dataset`
+#'
+#' @return Number of rows with duplicates preventing pivoting wider
+#' @export
+
+check_pivot_wider <- function(dataset) {
+
+  duplicates <- dataset$traits %>%
+    select(
+      dplyr::all_of(c("dataset_id", "trait_name", "value", "observation_id", "value_type",
+      "repeat_measurements_id", "method_id", "method_context_id"))
+    ) %>%
+    tidyr::pivot_wider(names_from = "trait_name", values_from = "value", values_fn = length) %>%
+    tidyr::pivot_longer(cols = 7:ncol(.)) %>%
+    dplyr::rename(dplyr::all_of(c("trait_name" = "name", "number_of_duplicates" = "value"))) %>%
+    select(
+      dplyr::all_of(c("dataset_id", "trait_name", "number_of_duplicates", "observation_id",
+      "value_type")), everything()
+    ) %>%
+    filter(.data$number_of_duplicates > 1) %>%
+    nrow()
+
+  if (duplicates == 0) {
+    invisible(TRUE)
+  } else {
+     invisible(FALSE)
+  }
+
+}
+
+
+#' @title Pivot long format data into a wide format
+#'
+#' @description `trait_pivot_wider` "widens" long format data ("tidy data").
+#'
+#' Databases built with `traits.build` are organised in a long format where observations are on different rows and the
+#' type of observation is denoted by various identifying columns (e.g `trait_name`, `dataset_id`,
+#' `observation_id`, etc.).
+#' This function converts the data into wide format so that each trait in its own column.
+#'
+#' @param traits The traits table from database (list object)
+#' @return A tibble in wide format
+#' @details `trait_pivot_wider` will return a single wide tibble; note that some meta-data columns
+#' (unit, replicates, measurement_remarks, basis_of_record, basis_of_value) will be excluded to
+#' produce a useful wide tibble.
+#' @examples
+#' \dontrun{
+#' data <- austraits$traits %>% filter(dataset_id == "Falster_2003")
+#' data # Long format
+#' traits_wide <- trait_pivot_wider(data)
+#' traits_wide # Wide format
+#' }
+#' @author Daniel Falster - daniel.falster@unsw.edu.au
+#' @export
+trait_pivot_wider <- function(traits) {
+
+  metadata_cols <- c("unit", "replicates", "measurement_remarks", "basis_of_value")
+
+  # A check for if there are more than 1 value_type for a given taxon_name, observation_id and method
+  check_value_type <- traits %>%
+    select(dplyr::all_of(c(
+      "trait_name", "value", "dataset_id", "observation_id", "method_id", "method_context_id",
+      "repeat_measurements_id", "value_type"))) %>%
+    dplyr::group_by(
+      .data$dataset_id, .data$observation_id, .data$method_id,
+      .data$method_context_id, .data$repeat_measurements_id) %>%
+    dplyr::summarise(n_value_type = length(unique(.data$value_type))) %>%
+    arrange(.data$observation_id) %>%
+    dplyr::filter(.data$n_value_type > 1)
+
+  if (nrow(check_value_type) > 1) {
+
+    traits %>%
+      select(-dplyr::all_of(metadata_cols)) %>%
+      # Sophie - what's the point of the group_by?
+      dplyr::group_by(
```

Comment:
Maybe because the ids aren't unique? so only makes sense to do dataset by dataset?

Reply:
The IDs should be unique I think. I tried omitting the `group_by` and didn't see anything different with the output, but would like someone else to try and verify.

Reply:
I think what's going on is that if you're maintaining columns that aren't the pivot, names_from or values_from columns you need to identify the pivot columns with group_by - but because now we're removing all the others there won't be a difference here.

Reply:
I tried without the `group_by` and it has the same number of columns with and without -- including columns that aren't needed for the pivot (redundant variables like `entity_type`, `population_id`, etc. The [only columns we remove](https://github.com/traitecoevo/traits.build/blob/24d9927f20128e7135e067b385c78f4e25d19eef/R/pivot.R#L78) are `c("unit", "replicates", "measurement_remarks", "basis_of_value")` right?